In [2]:
!apt-get moo

                 (__) 
                 (oo) 
           /------\/ 
          / |    ||   
         *  /\---/\ 
            ~~   ~~   
..."Have you mooed today?"...


В этом ноутбуке мы рассмотрим базовые команды командной строки в Linux

План
  - Команды UNIX
  - Файловая система
  - Аргументы к командам
  - Переменные окружения


# Команды в UNIX

Линукс принадлежит к большому семейству операционных систем под названием UNIX, которые происходят от ОС AT&T Unix, разработанной еще в 1970-х. Mac OS, кстати, тоже её потомок.

В операционных системах UNIX-типа для взаимодействия с ОС используют специальные команды. До появления привычного оконного интерфейса пользователи взаимодействовали с компьютером исключительно с помощью этих команд. 

Сейчас этот способ тоже используют, т. к. у него есть ряд преимуществ:
 - намного более удобней при подключение к компьютеру по сети;
 - иногда удобнее запустить одну команду, чем бродить по менюшкам в поисках нужной опции;
 - из команды можно составлять скрипты для автоматизации задач.



Небольшой ликбез по понятиям:
- **Командная строка** - интерфейс пользователя, окно для вывода системных сообщений и приёма команд;
- **Терминал** - программа, которая отображает строку. Например: *gnome-terminal*, *xterm*, *tilda*;
- **Оболочка (shell)** - программа которая выполняет команды. Например: *bash*, *zsh*, *sh*;
- **Консоль** - комплект устройств интерактивного ввода-вывода (дисплей, клавиатура, мышь) для управления компьютером.

Запускать команды можно как из терминала, так и из юпитера или колаба (Google Colab), поставив перед командой "!":

In [3]:
!echo "Hello World"

Hello World


# Файловая система

В Unix-системах все файлы образуют структуру похожую на дерево, в каждой директории (папке) хранятся файлы и другие директории. Самая первая директория которая содержит все остальные, называется корневой (*root*)


У каждой директории есть путь (по сути, просто строка) который показывает где она находится в этом дереве. "/" - это путь корневой директории. 

---



![alt text](http://linuxcommand.org/images/file_manager.jpg)

В командной строке все команды выполняются в какой-либо директории.

Представьте, что дерево файловой системы - это лабиринт, и каждая директория - это комната с файлами и дверьми в другие директории. В каждый момент времени вы находитесь в одной комнате.

Посмотрим, как ходить по нашему лабиринту.

In [10]:
# Выведем путь текущей директории
# pwd is for "print working directory"
!pwd 

/content


In [11]:
# Посмотрим, что в нашей директории есть
!ls 
#ls is for "list"

sample_data


In [12]:
#Создадим пустой файлик
!touch test-file
# пробел в имени - плохая идея, создадутся два файла
!touch hello world

# посмотрим что получилось
!ls

hello  sample_data  test-file  world


In [14]:
#Удаление файлов  ReMove
!rm world
!ls

rm: cannot remove 'world': No such file or directory
hello  sample_data  test-file


In [15]:
# Новая директория
!mkdir please_come_here


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если знаешь, что делаешь.

Это так называемые скрытые директории (hidden или dot directories/files) - их имя начинается с точки.


In [16]:
# Покажем больше информации и отобразим скрытые файлы
!ls -al
# Первое слово - команда, а второе - аргумент

total 20
drwxr-xr-x 1 root root 4096 Sep 15 14:25 .
drwxr-xr-x 1 root root 4096 Sep 15 14:06 ..
drwxr-xr-x 4 root root 4096 Sep  9 13:21 .config
-rw-r--r-- 1 root root    0 Sep 15 14:14 hello
drwxr-xr-x 2 root root 4096 Sep 15 14:25 please_come_here
drwxr-xr-x 1 root root 4096 Sep  9 13:22 sample_data
-rw-r--r-- 1 root root    0 Sep 15 14:14 test-file




```
.  - это путь к той директории в которой мы уже находимся
.. - это путь к родительской директории
.config - это скрытая папка
```

Каждая строчка - это одна из "дверей" в другую директорию либо файл.
*Относительный* путь это не полный "адрес" директории а то, куда нам нужно пойти из места в котором мы сейчас находимся. 


```
 drwxr-xr-x 1 root root 4096 Feb 15 17:21 please_come_here
 |                           дата создания    |
 |                                         _относительный_ путь
 |
  \ d - эта метка значит, что это путь к другой директории (d for directory)
```



In [17]:
# переход в другую директорию команда cd
!pwd
!cd please_come_here
!pwd

/content
/content


Заработало?

Не должно. Тут проблема в том что юпитер при исполнении каждой команды запускает новую оболочку в своей директории (той, откуда его вызвали), и после завершения команды она закрывается. Т. е. в одной оболочке он выполнил команду pwd, закрыл её, в другой выполнил команду cd please_come_here, закрыл её, и в третьей выполнил команду pwd и закрыл её.

Чтоб команды запускались в новой директории нужно поменять текущую директорию самого юпитера с помощью команды:

```
%cd
```


In [18]:
!pwd
%cd please_come_here
!pwd

/content
/content/please_come_here
/content/please_come_here


Если выполнить команду cd без аргументов - он вернёт вас в домашнюю директорию:

In [19]:
%cd
!pwd

/root
/root


Когда мы переходим в другую директорию, ее имя добавляется к нашему пути через "/".

Например, были в `/home`, перешли в директорию `Pictures `оказались в `/home/Pictures`.

Абсолютный путь файла в этой директории будет `/home/Picutes/image.png`. 

Если мы хотим взять файл из директории, которая является дочерней к той директории, где мы сейчас находимся, мы можем указать относительный путь файла.

Например, из директории `/home` мы можем обратиться к файлу по пути `Pictures/image.png`.

In [20]:
# Можно выполнить несколько команд последовательно:
# с помощью символа ; - для безусловного последовательного выполнения команд, они выполнятся последовательно одна за другой:
!pwd; cd please_come_here; pwd

# с помощью символа && - при этом вторая (и последующая) команда выполнится только в случае успешного выполнения первой (предыдущей):
!pwd && cd please_come_here && pwd

# с помощью символа || - при этом вторая (и последующая) команда выполнится только в случае неуспешного выполнения первой (предыдущей):
!pwd || cd please_come_here && pwd


/root
/bin/bash: line 0: cd: please_come_here: No such file or directory
/root
/root
/bin/bash: line 0: cd: please_come_here: No such file or directory
/root
/root


## Запись и чтение в файлы

In [21]:
# Вывести текст в консоль

!echo "Sample text Пример текста"
# Echo - потому что команда выводит на выход то же, что и получает

Sample text Пример текста


In [22]:
# Вместо консоли отправить в файл

!echo "London is the capital of Great Britain" >> English.txt


In [23]:
# Прочитать файл и вывести в консоль

!cat English.txt

London is the capital of Great Britain


In [25]:
!pwd

/root


In [27]:
!ls -al

total 72
drwx------ 1 root root 4096 Sep 15 14:33 .
drwxr-xr-x 1 root root 4096 Sep 15 14:06 ..
-r-xr-xr-x 1 root root 1169 Jan  1  2000 .bashrc
drwxr-xr-x 1 root root 4096 Sep 13 13:21 .cache
drwx------ 4 root root 4096 Sep 13 13:20 .config
-rw-r--r-- 1 root root   39 Sep 15 14:32 English.txt
-rw-r--r-- 1 root root 1256 Sep 15 14:33 example-com.html
drwxr-xr-x 1 root root 4096 Sep 13 13:20 .ipython
drwx------ 2 root root 4096 Sep 13 13:20 .jupyter
drwxr-xr-x 2 root root 4096 Sep  9 13:33 .keras
drwxr-xr-x 1 root root 4096 Sep 13 13:20 .local
drwxr-xr-x 4 root root 4096 Sep 13 13:20 .npm
-rw-r--r-- 1 root root  148 Aug 17  2015 .profile
-r-xr-xr-x 1 root root  254 Jan  1  2000 .tmux.conf
-rw-r--r-- 1 root root  165 Sep 13 13:20 .wget-hsts


## Скачивание из интернета

In [28]:
# сохраняем страницу по адресу 
#curl  client URL  >>  to file

!curl example.com >> example-com.html
!ls
!cat example-com.html


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1256  100  1256    0     0   7753      0 --:--:-- --:--:-- --:--:--  7753
English.txt  example-com.html
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:vis

In [29]:
!curl instagram.com

Иногда сервера пытаются перенаправить запрос на другой адрес. Curl не следует по таким перенаправлениям. Есть програмка wget которая может переходить по таким ссылкам и вообще проходить по всем ссылкам на странице и скачивать их.

In [30]:
!wget instagram.com -O insta.html

--2022-09-15 14:41:58--  http://instagram.com/
Resolving instagram.com (instagram.com)... 157.240.247.174, 2a03:2880:f273:e9:face:b00c:0:4420
Connecting to instagram.com (instagram.com)|157.240.247.174|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://instagram.com/ [following]
--2022-09-15 14:41:58--  https://instagram.com/
Connecting to instagram.com (instagram.com)|157.240.247.174|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://www.instagram.com/ [following]
--2022-09-15 14:41:58--  https://www.instagram.com/
Resolving www.instagram.com (www.instagram.com)... 157.240.201.174, 2a03:2880:f273:e9:face:b00c:0:4420
Connecting to www.instagram.com (www.instagram.com)|157.240.201.174|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 104653 (102K) [text/html]
Saving to: ‘insta.html’

insta.html          100%[===================>] 102.20K  --.-KB/s    in 0.08s   

2022-0

Обратите внимание на

`HTTP request sent, awaiting response... 301 Moved Permanently` (четвёртая строка)


# Аргументы и параметры к командам

Обычно **аргументы** – это то, над чем выполняется команда. Например, команда перехода в другую директорию должна получить в качестве аргумента путь директории, в которую следует перейти. Команде вывода на экран содержимого файла надо передать имя файла.

**Параметры** модифицируют, как бы видоизменяют/управляют работой программы. Например, с помощью команды просмотра содержимого каталога можно вывести простой список вложенных объектов. Используя же определенный параметр, можно получить подробные сведения о каждом файле и папке.

Аргументы указывают через пробел, как отдельные слова.

Параметры указывают после дефиса: `-f` (однобуквенные) либо после двух дефисов: `--file`.

In [32]:
# календарик на весь год

!ncal 2022


                                  2022
    January           February          March             April             
Su     2  9 16 23 30     6 13 20 27        6 13 20 27        3 10 17 24   
Mo     3 10 17 24 31     7 14 21 28        7 14 21 28        4 11 18 25   
Tu     4 11 18 25     1  8 15 22        1  8 15 22 29        5 12 19 26   
We     5 12 19 26     2  9 16 23        2  9 16 23 30        6 13 20 27   
Th     6 13 20 27     3 10 17 24        3 10 17 24 31        7 14 21 28   
Fr     7 14 21 28     4 11 18 25        4 11 18 25        1  8 15 22 29   
Sa  1  8 15 22 29     5 12 19 26        5 12 19 26        2  9 16 23 30   

    May               June              July              August            
Su  1  8 15 22 29        5 12 19 26        3 10 17 24 31     7 14 21 28   
Mo  2  9 16 23 30        6 13 20 27        4 11 18 25     1  8 15 22 29   
Tu  3 10 17 24 31        7 14 21 28        5 12 19 26     2  9 16 23 30   
We  4 11 18 25        1  8 15 22 29        6 13 20 27   

In [33]:
# передаем 2 аргумента

!ncal 09 2020


    September 2020    
Su     6 13 20 27   
Mo     7 14 21 28   
Tu  1  8 15 22 29   
We  2  9 16 23 30   
Th  3 10 17 24      
Fr  4 11 18 25      
Sa  5 12 19 26      


In [34]:
# передаем параметр -b

!ncal -b 09 2020


   September 2020     
Su Mo Tu We Th Fr Sa  
       1  2  3  4  5  
 6  7  8  9 10 11 12  
13 14 15 16 17 18 19  
20 21 22 23 24 25 26  
27 28 29 30           
                      


In [35]:
# Однобуквенные параметры можно склеивать

!ncal -b -w 09 2020
!ncal -bw 09 2020

     September 2020        
 w| Su Mo Tu We Th Fr Sa   
35|        1  2  3  4  5   
36|  6  7  8  9 10 11 12   
37| 13 14 15 16 17 18 19   
38| 20 21 22 23 24 25 26   
39| 27 28 29 30            
                           
     September 2020        
 w| Su Mo Tu We Th Fr Sa   
35|        1  2  3  4  5   
36|  6  7  8  9 10 11 12   
37| 13 14 15 16 17 18 19   
38| 20 21 22 23 24 25 26   
39| 27 28 29 30            
                           


# Переменные окружения

Оболочка в себе содержит так называемые переменные окружения - глобальные константы которые доступны всем программам которые она выполняет. В них удобно хранить например путь к конфигурационным файлам или порт какого нибудь сервиса.

Переменные окружения зачастую определяют как ведет себя программа. Например, Python ищет пакеты в переменной `PYTHONPATH`.

В переменной `PATH` находятся пути в которых Bash ищет исполняемые файлы других программ.


In [36]:
# объявляем переменную и считываем ее ($ перед именем)
!FOO=kindersurprise && echo $FOO

kindersurprise


In [37]:
!echo $PATH
!echo $PYTHONPATH

/opt/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin
/env/python


Мы можем создать новую папку и заставить питон думать что в ней лежит модуль:

In [39]:
!rm -r fakemodule
!mkdir fakemodule
!mkdir fakemodule/power
!echo "print('A great power comes with a great responsibility')" >> fakemodule/power/__init__.py


In [44]:
!ls fakemodule/power/
!touch fakemodule/power/__init__.py

__init__.py  __pycache__


In [56]:
%cd ..

/root


In [65]:
!ls

English.txt  example-com.html  fakemodule  insta.html


In [60]:
!export a="Эта переменная будет доступна другим программам вызванным в той же оболочке" && echo $a
# Но тут уже не получится, для каждой команды Jupyter запускает свою оболочку 
!echo $a

Эта переменная будет доступна другим программам вызванным в той же оболочке



In [59]:
!PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule && python -c 'import power'

A great power comes with a great responsibility


Конструкция `$(pwd)` подставляет вывод команды `pwd` в выражение. 

```
PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule

```
После подстановки переменных мы получим:

```
PYTHONPATH=<то что было в PYTHONPATH>:<текущая директория>/fakemodule

```
Это добавит нашу fakemodule в переменную окружения и позволит питону найти наш модуль.


In [67]:
!echo $(pwd)

/root


С помощью этой конструкции можно из результата выполнения команды создать переменную и сразу ей воспользоваться:

In [71]:
!dt_now=$(date); echo $dt_now

Thu Sep 15 15:37:18 UTC 2022


Для более углублённого изучения Linux в целом и командной оболочки в частности рекомендуем пройти курс Введение в Linux от Института биоинформатики https://stepik.org/course/73

In [2]:
!pip install python-slugify

from slugify import slugify



  Using cached python_slugify-8.0.1-py2.py3-none-any.whl (9.7 kB)
  Using cached text_unidecode-1.3-py2.py3-none-any.whl (78 kB)

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: pip install --upgrade pip


In [3]:
filename = "Пример документа"
slugified_filename = slugify(filename, lowercase=False, separator="_")
print(slugified_filename)  # Пример_документа

Primer_dokumenta


In [ ]:
!